utils/swift_build_support/swift_build_support/products/product.py

```diff
@@ -0,0 +1,32 @@
+# swift_build_support/products/product.py -----------------------*- python -*-
+#
+# This source file is part of the Swift.org open source project
+#
+# Copyright (c) 2014 - 2016 Apple Inc. and the Swift project authors
+# Licensed under Apache License v2.0 with Runtime Library Exception
+#
+# See http://swift.org/LICENSE.txt for license information
+# See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+#
+# ----------------------------------------------------------------------------
+
+
+class Product(object):
+    @classmethod
+    def product_name(klass):
+        """product_name() -> str
+
+        The identifier-style name to use for this product.
+        """
+        return klass.__name__.lower()
+
+    @classmethod
+    def get_build_directory_name(klass, host_target):
+        return "{}-{}".format(klass.product_name(),
+                              host_target.name)
+
+    def __init__(self, args, toolchain, source_dir, build_dir):
+        self.args = args
```

Comment:
I agree, I don't like this either parsing of arguments outside of the main `build-script`.

I think it would be better to keep the common initializer, though, and instead refactor it to take a more explicit value type instead of the full argument set.